### ISSUE TITLE
GLMakie precompile hangs on Windows 10, Julia 1.9

### ISSUE THREAD

**Issue by Ininterrompue** [JuliaLang/julia#49743]
Putting https://github.com/MakieOrg/Makie.jl/issues/2941 here:

After updating to Julia 1.9, the following is shown in the REPL after `]up GLMakie`:

```
Precompiling project...
  Progress [>                                        ]  0/1
  ◒ GLMakie
```

After a bit, CPU usage hovers around 0%. CairoMakie precompiles fine. GLMakie precompiles fine on 1.8.5. 

Attempted reinstalling GLMakie, restarting Julia, rebooting computer, using juliaup. Using a 64-bit system.

**Comment by vtjnash** [comment]
Can you attach to the process with gdb and get any sort of backtrace or other info on its status?

**Comment by ViralBShah** [comment]
cc @SimonDanisch

**Comment by SimonDanisch** [comment]
I can't reproduce this on my windows 11 PC... Unless someone can show this only hangs in the newest Makie releases, I'm assuming it's a Julia issue.
```julia
(@v1.9) pkg> activate --temp; add GLMakie
...
Precompiling project...
  10 dependencies successfully precompiled in 175 seconds. 194 already precompiled.
```
```julia

julia> versioninfo()
Julia Version 1.9.0
Commit 8e63055292 (2023-05-07 11:25 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: 16 × AMD Ryzen 7 5800H with Radeon Graphics
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-14.0.6 (ORCJIT, znver3)
  Threads: 1 on 16 virtual cores
Environment:
  JULIA_IMAGE_THREADS = 1
  JULIA_PKG_PRECOMPILE_AUTO = 1
```

I just noticed that I have `JULIA_IMAGE_THREADS = 1` so I tried with `JULIA_IMAGE_THREADS = 8`, which also compiles fine...
Can someone check if disabling the precompile workload helps?
```julia
using Preferences
set_preferences!(Base.UUID( "e9467ef8-e4e7-5192-8a1a-b1aee30e663a), "precompile_workload" => false; force=true)
```

**Comment by matthijscox** [comment]
Me and a few other colleagues have the same problem with GLMakie on Julia 1.9 on Windows 10. However, some other colleagues with similar laptops do not have this problem. 

I did the following:
* deleted my entire `.julia/compiled/v1.9` folder, 
* started Julia 1.9, 
* activated a temporary environment, 
* added GLMakie, 
* `using GLMakie`
* watching the `.julia/compiled/v1.9` folder fill with packages

It spent most time on `Makie` itself (about 2 minutes), but got past that. The last package to be added to the folder was `PNGFiles`, after that my Julia CPU usage dropped to 0%, but it's still waiting on the REPL. In total all that took about 8 minutes.

For the last 10 minutes nothing has happened.

I killed the precompilation process with `ctrl+c`. I don't know if this means anything, but I get the following error
```julia
ERROR: InterruptException:
Stacktrace:
  [1] try_yieldto(undo::typeof(Base.ensure_rescheduled))
    @ Base .\task.jl:920
  [2] wait()
    @ Base .\task.jl:984
  [3] wait(c::Base.GenericCondition{Base.Threads.SpinLock}; first::Bool)
    @ Base .\condition.jl:130
  [4] wait
    @ .\condition.jl:125 [inlined]
  [5] wait(x::Base.Process)
    @ Base .\process.jl:661
  [6] success(x::Base.Process)
    @ Base .\process.jl:523
  [7] compilecache(pkg::Base.PkgId, path::String, internal_stderr::IO, internal_stdout::IO, keep_loaded_modules::Bool)
    @ Base .\loading.jl:2163
  [8] compilecache
    @ .\loading.jl:2127 [inlined]
  [9] _require(pkg::Base.PkgId, env::String)
    @ Base .\loading.jl:1770
 [10] _require_prelocked(uuidkey::Base.PkgId, env::String)
    @ Base .\loading.jl:1625
 [11] macro expansion
    @ .\loading.jl:1613 [inlined]
 [12] macro expansion
    @ .\lock.jl:267 [inlined]
 [13] require(into::Module, mod::Symbol)
    @ Base .\loading.jl:1576
```

I then ran the command from Simon:
```
using Preferences
set_preferences!(Base.UUID("e9467ef8-e4e7-5192-8a1a-b1aee30e663a"), "precompile_workload" => false; force=true)
using GLMakie
```

Oh, it continued! Trying a plot... it's working!

That's nice. But what does this mean?

**Comment by matthijscox** [comment]
Alright, so for now I'll have to keep the precompile workload disabled, until someone figures out a fix?

Is there something that could help pinpoint the root cause?

**Comment by ultimatile** [comment]
I am experiencing the same issue with macOS Ventura on M1 Mac.
The strange thing is that it doesn't happen on my MacBook, but it does on Mac Studio.
Of course, they are not exactly the same.
Please let me know if I should include details of the environment in which it occurs, or if I should separate the issue.

**Comment by SimonDanisch** [comment]
@Moelf posted on slack, that Makie itself also hangs compiling, so any weird opengl / driver interaction can be excluded, since Makie is pretty much pure Julia.

**Comment by Moelf** [comment]
(still a chance I'm observing something different)

**Comment by IanButterworth** [comment]
You should be able to get a profile on the precompiling process via sending it a siginfo/sigusr1. I'd recommend doing that to the code load precompiler rather than Pkg, as you should get the output right away.

**Comment by Sagnac** [comment]
I experienced the same issue and can also confirm that disabling the precompile workload solved it.

**Comment by gbaraldi** [comment]
Can you share the full environment and also what graphics cards are you using? (intel, nvidia etc)

**Comment by Ininterrompue** [comment]
I attempted this on two more Windows 10 computers and both of them succeeded at precompiling GLMakie on 1.9.

The one that hangs for me has the following specs:

CPU: 12 × Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
GPU: Nvidia GeForce GTX 1660 Ti

This problem continues to occur even after updating to the latest version of Windows 10 and updating GPU drivers.

**Comment by Ininterrompue** [comment]
On 1.9.1, GLMakie precompiles successfully for me on the computer that was giving me issues.

**Comment by ufechner7** [comment]
On 1.9.1 on Windows package compiler hangs for me, I am using Makie:

```
[ Info: Loading packages ...
[ Info: Creating sysimage ...
Precompiling project...
  65 dependencies successfully precompiled in 73 seconds. 181 already precompiled.
[ Info: PackageCompiler: Executing C:\Users\uwefechner\repos\KiteModels.jl\test\test_for_precompile.jl => C:\Users\UWEFEC~1\AppData\Local\Temp\jl_packagecompiler_zB7cFs\jl_E87A.tmp
[ Info: Precompile script has completed execution.
[ Info: PackageCompiler: Done
✖ [01h:00m:28s] PackageCompiler: compiling incremental system image
ERROR: LoadError: failed process: Process(`'C:\Users\uwefechner\.julia\juliaup\julia-1.9.1+0.x64.w64.mingw32\bin\julia.exe' --color=yes --startup-file=no --pkgimages=no --cpu-target=native '--sysimage=C:\Users\uwefechner\.julia\juliaup\julia-1.9.1+0.x64.w64.mingw32\lib\julia\sys.dll' '--project=C:\Users\UWEFEC~1\AppData\Local\Temp\jl_W8rPbw' '--output-o=C:\Users\UWEFEC~1\AppData\Local\Temp\jl_ZOzQvEbGBa.o' 'C:\Users\UWEFEC~1\AppData\Local\Temp\jl_D4jsvovqbk'`, ProcessExited(130)) [130]

Stacktrace:
  [1] pipeline_error
    @ .\process.jl:565 [inlined]
  [2] run(::Cmd; wait::Bool)
    @ Base .\process.jl:480
  [3] run
    @ .\process.jl:477 [inlined]
  [4] #20
    @ C:\Users\uwefechner\.julia\packages\PackageCompiler\b2smD\ext\TerminalSpinners.jl:157 [inlined]
  [5] spin(f::PackageCompiler.var"#20#22"{Cmd}, s::PackageCompiler.TerminalSpinners.Spinner{Base.TTY})
    @ PackageCompiler.TerminalSpinners C:\Users\uwefechner\.julia\packages\PackageCompiler\b2smD\ext\TerminalSpinners.jl:164
  [6] macro expansion
    @ C:\Users\uwefechner\.julia\packages\PackageCompiler\b2smD\ext\TerminalSpinners.jl:157 [inlined]
  [7] create_sysimg_object_file(object_file::String, packages::Vector{String}, packages_sysimg::Set{Base.PkgId}; project::String, base_sysimage::String, precompile_execution_file::Vector{String}, precompile_statements_file::Vector{String}, cpu_target::String, script::Nothing, sysimage_build_args::Cmd, extra_precompiles::String, incremental::Bool)
    @ PackageCompiler C:\Users\uwefechner\.julia\packages\PackageCompiler\b2smD\src\PackageCompiler.jl:128
  [8] create_sysimg_object_file
    @ C:\Users\uwefechner\.julia\packages\PackageCompiler\b2smD\src\PackageCompiler.jl:290 [inlined]
  [9] create_sysimage(packages::Vector{Symbol}; sysimage_path::String, project::String, precompile_execution_file::String, precompile_statements_file::Vector{String}, incremental::Bool, filter_stdlibs::Bool, cpu_target::String, script::Nothing, sysimage_build_args::Cmd, include_transitive_dependencies::Bool, base_sysimage::Nothing, julia_init_c_file::Nothing, version::Nothing, soname::Nothing, compat_level::String, extra_precompiles::String)
    @ PackageCompiler C:\Users\uwefechner\.julia\packages\PackageCompiler\b2smD\src\PackageCompiler.jl:586
 [10] top-level scope
    @ C:\Users\uwefechner\repos\KiteModels.jl\test\create_sys_image.jl:13
 [11] include(fname::String)
    @ Base.MainInclude .\client.jl:478
 [12] top-level scope
    @ none:1
in expression starting at C:\Users\uwefechner\repos\KiteModels.jl\test\create_sys_image.jl:13

real    66m30.541s
user    0m0.045s
sys     0m0.243s
(base)

```

**Comment by KristofferC** [comment]
That seems different sincs the precompilation finished.

**Comment by SimonDanisch** [comment]
Can everyone who run into this please post the OS, GPU and CPU or anything other that could be special about the system?
I've tried lots of things, but I can't reproduce this on neither my 2 windows PCs nor on a linux machine.

**Comment by ufechner7** [comment]
> That seems different sincs the precompilation finished.

OK, precompilation finished, but creating a system image did not finish. I pressed <ctrl><c>.

**Comment by ViralBShah** [comment]
If anyone here is able to reproduce reliably on 1.9, an RR trace would be appreciated.

**Comment by olejorik** [comment]
I've been able to nail it down. In my case it _is_ related to the GPU drivers. If I start julia with my external monitor not connected to the laptop, it runs on Intel GPU, and I can precompile GLMakie successfully, and it works just fine. 
![image](https://github.com/JuliaLang/julia/assets/12269930/f49361c5-f63e-46f2-830e-31d82d296632)


If I connect the external monitor, NVIDIA GPU becomes active, and GLMakie fails to draw its window.
![image](https://github.com/JuliaLang/julia/assets/12269930/ccb11f6e-f57f-4c30-bb3b-7972696b380d) 
![image](https://github.com/JuliaLang/julia/assets/12269930/cfc0be9b-5a1d-4cb4-934b-4d98a3fa6521)

Precompilation with the attached external monitor also hangs. I've tried to roll back the NVIDIA driver (which was recently updated), but it didn't change anything.

My system runs on Windows 11,

Julia Version 1.9.1
Commit 147bdf428c (2023-06-07 08:27 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: 8 × Intel(R) Core(TM) i7-6820HQ CPU @ 2.70GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-14.0.6 (ORCJIT, skylake)
  Threads: 1 on 8 virtual cores
  
 GPU1: Intel HD Graphics 530, Driver 30.0.101.1338
 GPU2: NVIDIA Quadro M2000M, Diver 31.0.15.2879

**Comment by SimonDanisch** [comment]
What does `using GLFW; GLFW.Window()` do in the case it hangs?

**Comment by SimonDanisch** [comment]
That sounds like it works (closing is something that needs to be manually implemented^^).

**Comment by matthijscox** [comment]
I can confirm that on my side GLMakie is now working on 1.9.1. Quite happy with that :)

I'll check with a few colleagues as well.

I have NOT tried the PackageCompiler route. That sounds less critical to me, though it's good to know if we can expect any issues there, in case someone wants to deploy standalone Makie apps.

**Comment by ViralBShah** [comment]
Thanks - it makes sense to file at Plots.jl. Perhaps we can close this for now. If people still are facing this going forward, we can always reopen.

**Comment by KristofferC** [comment]
https://github.com/JuliaLang/julia/issues/49743#issuecomment-1584253821 still seems like it is valid

**Comment by ViralBShah** [comment]
I thought @SimonDanisch 's response suggested this is not a Julia compilation issue. Please reopen if there is still a valid bug in here, or perhaps a new issue.

**Comment by SimonDanisch** [comment]
No, sorry ! I just wanted to make sure that glmakie is able to open a window without hanging, under the problematic condition! That it works points further towards a compilation specific issue !

**Comment by olejorik** [comment]
Today, after updating to julia 1.9.2, the [issue](https://github.com/JuliaLang/julia/issues/49743#issuecomment-1584253821) I reported one month ago is gone! The precompilation took about 8 minutes, but everything works nicely now :-) 
Thanks to everyone!
